File: plugin/python/larch/implement/self_edit_log.py
"""Self-edit attribution log for ``/implement`` (issue #6876).

File-mutating subprocesses spawned by ``/implement`` — the ``checks repair-loop``
lint-fix tiers and the Step 3 pre-commit ruff autofix — edit tracked files in
place. Those edits carry no attribution the orchestrator can consult, so an
observed between-action file change can be misattributed to a concurrent or
external runner, producing a false alarm and an unnecessary operator halt.

Each mutating subprocess records the repo-relative paths it changed here (one
row per path: recorded epoch seconds, source, path, post-edit sha256). The
orchestrator consults the log via ``checks self-edit-log`` before ever
concluding that a working-tree change came from another runner: process
introspection (``ps``) and ``stat`` mtime cannot attribute an edit once the
spawning subprocess has exited, so the log is the authority.

Stdlib-only. Writes land under ``$IMPLEMENT_TMPDIR`` and are consumed by the same
session, so paths are stored verbatim (not routed through ``redact.redact``,
which would strip the tmpdir path literals the orchestrator needs to match).
"""

from __future__ import annotations

import contextlib
import hashlib
import re
import time
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
from typing import Final

SELF_EDIT_LOG_NAME: Final = "self-edit-log.tsv"
_FIELDS: Final = ("recorded_epoch_s", "source", "path", "post_sha256")
_HEADER_LINE: Final = "\t".join(_FIELDS) + "\n"
_FIELD_COUNT: Final = len(_FIELDS)
_SOURCE_MAX_LEN: Final = 64
_SOURCE_SANITIZE_RE: Final = re.compile(r"[^A-Za-z0-9_.:@/-]")


@dataclass(frozen=True)
class SelfEditRecord:
    recorded_epoch_s: int
    source: str
    path: str
    post_sha256: str


def self_edit_log_path(tmpdir: str | Path) -> Path:
    return Path(tmpdir) / SELF_EDIT_LOG_NAME


def normalize_path(value: str) -> str:
    """Strip TSV-hostile control characters so a path round-trips exactly.

    Applied to every recorded path and to a ``show --path`` query so the two
    match on the same normalized form.
    """
    return value.replace("\t", " ").replace("\r", " ").replace("\n", " ").strip()


def _sanitize_source(source: str) -> str:
    cleaned = _SOURCE_SANITIZE_RE.sub("-", source.strip())[:_SOURCE_MAX_LEN]
    return cleaned or "unknown"


def file_sha256(repo_root: str | Path, path: str) -> str:
    """sha256 of a repo-relative file, or a bounded sentinel when unreadable."""
    try:
        target = Path(repo_root) / path
        if not target.is_file() or target.is_symlink():
            return "missing"
        return hashlib.sha256(target.read_bytes()).hexdigest()
    except OSError:
        return "unreadable"


def digest_paths(repo_root: str | Path, paths: Iterable[str]) -> dict[str, str]:
    """Map each repo-relative path to its current sha256 (see ``file_sha256``)."""
    root = Path(repo_root)
    return {path: file_sha256(root, path) for path in paths}


def _append_rows(path: Path, rows: list[tuple[int, str, str, str]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.is_symlink():
        return
    needs_header = not path.exists() or path.stat().st_size == 0
    with path.open("a", encoding="utf-8") as handle:
        if needs_header:
            _ = handle.write(_HEADER_LINE)
        for epoch_s, source, rel_path, sha in rows:
            _ = handle.write(f"{epoch_s}\t{source}\t{rel_path}\t{sha}\n")
    with contextlib.suppress(OSError):
        path.chmod(0o600)


def record_self_edits(
    *,
    tmpdir: str | Path,
    source: str,
    paths: Iterable[str],
    repo_root: str | Path,
    now_epoch_s: int | None = None,
) -> int:
    """Append one attribution row per changed path. Returns the row count.

    Best-effort: never raises, so a checks/repair-loop caller is never disrupted
    by logging. Returns 0 (nothing recorded) when the tmpdir is unusable, no
    paths are supplied, or any write fails.
    """
    written = 0
    with contextlib.suppress(Exception):
        tmp = Path(tmpdir)
        if not tmp.is_dir() or tmp.is_symlink():
            return 0
        epoch_s = int(time.time()) if now_epoch_s is None else int(now_epoch_s)
        source_token = _sanitize_source(source)
        rows: list[tuple[int, str, str, str]] = []
        seen: set[str] = set()
        for raw in paths:
            rel = normalize_path(str(raw))
            if not rel or rel in seen:
                continue
            seen.add(rel)
            rows.append((epoch_s, source_token, rel, file_sha256(repo_root, rel)))
        if not rows:
            return 0
        _append_rows(self_edit_log_path(tmp), rows)
        written = len(rows)
    return written


def read_self_edits(tmpdir: str | Path) -> list[SelfEditRecord]:
    """Parse recorded attribution rows; tolerant of a missing/partial log."""
    path = self_edit_log_path(tmpdir)
    records: list[SelfEditRecord] = []
    try:
        if not path.is_file() or path.is_symlink():
            return []
        text = path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return []
    header = _HEADER_LINE.rstrip("\n")
    for line in text.splitlines():
        if not line or line == header:
            continue
        parts = line.split("\t")
        if len(parts) != _FIELD_COUNT:
            continue
        epoch_raw, source, rel_path, sha = parts
        try:
            epoch_s = int(epoch_raw)
        except ValueError:
            continue
        records.append(
            SelfEditRecord(recorded_epoch_s=epoch_s, source=source, path=rel_path, post_sha256=sha)
        )
    return records

File: crates/larch-core/src/implement/self_edit_log.rs
//! Self-edit attribution log reader for `/implement` (issue #6876).
//!
//! File-mutating subprocesses spawned by `/implement` record the repo-relative
//! paths they changed. The orchestrator consults this log before concluding
//! that a between-action working-tree change came from another runner.

use std::{
    collections::HashMap,
    env, fs,
    io::Write as _,
    path::{Path, PathBuf},
    time::{SystemTime, UNIX_EPOCH},
};

use sha2::{Digest as _, Sha256};

/// File name of the attribution log inside an implement tmpdir.
pub const SELF_EDIT_LOG_NAME: &str = "self-edit-log.tsv";

const HEADER: &str = "recorded_epoch_s\tsource\tpath\tpost_sha256";
const HEADER_LINE: &str = "recorded_epoch_s\tsource\tpath\tpost_sha256\n";
const SESSION_PREFIXES: [&str; 2] = ["claude-implement-", "claude-review-"];
const SOURCE_MAX_LEN: usize = 64;

/// One recorded self-edit attribution row.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SelfEditRecord {
    /// Epoch seconds the row was recorded.
    pub recorded_epoch_s: i64,
    /// Sanitized source token of the mutating subprocess.
    pub source: String,
    /// Repository-relative path that was changed.
    pub path: String,
    /// SHA-256 of the file right after the edit.
    pub post_sha256: String,
}

/// Strip TSV-hostile control characters so a path round-trips exactly.
#[must_use]
pub fn normalize_path(value: &str) -> String {
    value.replace(['\t', '\r', '\n'], " ").trim().to_owned()
}

/// Parse recorded attribution rows; tolerant of a missing or partial log.
#[must_use]
pub fn read_self_edits(tmpdir: &Path) -> Vec<SelfEditRecord> {
    let path = tmpdir.join(SELF_EDIT_LOG_NAME);
    let Ok(metadata) = fs::symlink_metadata(&path) else {
        return Vec::new();
    };
    if metadata.file_type().is_symlink() || !metadata.is_file() {
        return Vec::new();
    }
    let Ok(bytes) = fs::read(&path) else {
        return Vec::new();
    };
    let text = String::from_utf8_lossy(&bytes);
    let mut records = Vec::new();
    for line in text.lines() {
        if line.is_empty() || line == HEADER {
            continue;
        }
        let fields: Vec<&str> = line.split('\t').collect();
        let [epoch_raw, source, rel_path, sha] = fields.as_slice() else {
            continue;
        };
        let Ok(recorded_epoch_s) = epoch_raw.trim().parse::<i64>() else {
            continue;
        };
        records.push(SelfEditRecord {
            recorded_epoch_s,
            source: (*source).to_owned(),
            path: (*rel_path).to_owned(),
            post_sha256: (*sha).to_owned(),
        });
    }
    records
}

/// SHA-256 of a repo-relative file, or a bounded sentinel when unreadable.
#[must_use]
pub fn file_sha256(repo_root: &Path, rel: &str) -> String {
    let target = repo_root.join(rel);
    match fs::symlink_metadata(&target) {
        Ok(metadata) if metadata.is_file() && !metadata.file_type().is_symlink() => {
            fs::read(&target).map_or_else(
                |_| "unreadable".to_owned(),
                |bytes| format!("{:x}", Sha256::digest(bytes)),
            )
        }
        Ok(_) | Err(_) => "missing".to_owned(),
    }
}

/// Map each repo-relative path to its current sha256 (see [`file_sha256`]).
#[must_use]
pub fn digest_paths(repo_root: &Path, paths: &[String]) -> HashMap<String, String> {
    paths
        .iter()
        .map(|path| (path.clone(), file_sha256(repo_root, path)))
        .collect()
}

/// Sanitize a source token to the `[A-Za-z0-9_.:@/-]` grammar, bounded length.
fn sanitize_source(source: &str) -> String {
    let cleaned: String = source
        .trim()
        .chars()
        .map(|ch| {
            if ch.is_ascii_alphanumeric() || matches!(ch, '_' | '.' | ':' | '@' | '/' | '-') {
                ch
            } else {
                '-'
            }
        })
        .take(SOURCE_MAX_LEN)
        .collect();
    if cleaned.is_empty() {
        "unknown".to_owned()
    } else {
        cleaned
    }
}

fn append_rows(path: &Path, rows: &[(i64, String, String, String)]) {
    if let Some(parent) = path.parent() {
        let _ignored = fs::create_dir_all(parent);
    }
    if fs::symlink_metadata(path).is_ok_and(|meta| meta.file_type().is_symlink()) {
        return;
    }
    let needs_header = fs::metadata(path).map_or(true, |meta| meta.len() == 0);
    let Ok(mut handle) = fs::OpenOptions::new().create(true).append(true).open(path) else {
        return;
    };
    if needs_header {
        let _ignored = handle.write_all(HEADER_LINE.as_bytes());
    }
    for (epoch_s, source, rel_path, sha) in rows {
        let _ignored =
            handle.write_all(format!("{epoch_s}\t{source}\t{rel_path}\t{sha}\n").as_bytes());
    }
    drop(handle);
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt as _;
        let _ignored = fs::set_permissions(path, fs::Permissions::from_mode(0o600));
    }
}

/// Append one attribution row per changed path; returns the row count.
///
/// Best-effort: never panics, so a checks or repair-loop caller is never
/// disrupted by logging. Returns 0 when the tmpdir is unusable, no paths are
/// supplied, or any write fails. `now_epoch_s` overrides the clock for tests.
#[must_use]
pub fn record_self_edits(
    tmpdir: &Path,
    source: &str,
    paths: &[String],
    repo_root: &Path,
    now_epoch_s: Option<i64>,
) -> usize {
    let Ok(metadata) = fs::symlink_metadata(tmpdir) else {
        return 0;
    };
    if !metadata.is_dir() || metadata.file_type().is_symlink() {
        return 0;
    }
    let epoch_s = now_epoch_s.unwrap_or_else(|| {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_or(0, |elapsed| {
                i64::try_from(elapsed.as_secs()).unwrap_or(i64::MAX)
            })
    });
    let source_token = sanitize_source(source);
    let mut rows: Vec<(i64, String, String, String)> = Vec::new();
    let mut seen: Vec<String> = Vec::new();
    for raw in paths {
        let rel = normalize_path(raw);
        if rel.is_empty() || seen.contains(&rel) {
            continue;
        }
        seen.push(rel.clone());
        let sha = file_sha256(repo_root, &rel);
        rows.push((epoch_s, source_token.clone(), rel, sha));
    }
    if rows.is_empty() {
        return 0;
    }
    append_rows(&tmpdir.join(SELF_EDIT_LOG_NAME), &rows);
    rows.len()
}

fn canonical_dir(path: &Path) -> Option<PathBuf> {
    let metadata = fs::symlink_metadata(path).ok()?;
    if !metadata.is_dir() || metadata.file_type().is_symlink() {
        return None;
    }
    fs::canonicalize(path).ok()
}

/// Validate an implement or review session tmpdir under its accepted roots.
///
/// Mirrors the shared `checks` tmpdir contract: an absolute, non-symlink
/// session directory whose canonical name carries a session prefix and that
/// lives under the larch session cache or directly under the system tmpdir.
#[must_use]
pub fn validate_session_tmpdir(tmpdir: &str) -> Option<PathBuf> {
    if !tmpdir.starts_with('/') {
        return None;
    }
    let canonical = canonical_dir(Path::new(tmpdir))?;
    let name = canonical.file_name()?.to_string_lossy().into_owned();
    if !SESSION_PREFIXES
        .iter()
        .any(|prefix| name.starts_with(prefix))
    {
        return None;
    }
    let xdg_cache = env::var("XDG_CACHE_HOME").unwrap_or_default();
    let xdg_cache = xdg_cache.trim();
    let cache_root = if xdg_cache.is_empty() {
        PathBuf::from(env::var("HOME").unwrap_or_default()).join(".cache")
    } else {
        PathBuf::from(xdg_cache)
    };
    if let Some(sessions) = canonical_dir(&cache_root.join("larch").join("sessions"))
        && canonical.starts_with(&sessions)
    {
        return Some(canonical);
    }
    let parent = canonical.parent()?.to_path_buf();
    for candidate in ["/tmp", "/private/tmp"] {
        if canonical_dir(Path::new(candidate)).is_some_and(|resolved| parent == resolved) {
            return Some(canonical);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn normalize_path_strips_control_characters() {
        assert_eq!(normalize_path("  a\tb\rc\nd  "), "a b c d");
        assert_eq!(normalize_path(""), "");
    }

    #[test]
    fn read_self_edits_skips_header_and_malformed_rows() {
        let root = TempDir::new().expect("temp");
        fs::write(
            root.path().join(SELF_EDIT_LOG_NAME),
            format!("{HEADER}\n1\tlint-fix\ta.py\tdeadbeef\nshort\trow\n2\tsrc\tb.py\n3x\ts\tc.py\th\n7\ts2\td.py\th2\n"),
        )
        .expect("write");
        let records = read_self_edits(root.path());
        assert_eq!(records.len(), 2);
        assert_eq!(records[0].path, "a.py");
        assert_eq!(records[0].recorded_epoch_s, 1);
        assert_eq!(records[1].source, "s2");
        assert!(read_self_edits(&root.path().join("absent")).is_empty());
    }

    #[test]
    fn file_sha256_reports_sentinels() {
        let root = TempDir::new().expect("temp");
        fs::write(root.path().join("a.txt"), "x").expect("write");
        assert_eq!(
            file_sha256(root.path(), "a.txt"),
            "2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881"
        );
        assert_eq!(file_sha256(root.path(), "missing.txt"), "missing");
    }

    #[test]
    fn validate_session_tmpdir_requires_prefix_and_root() {
        assert!(validate_session_tmpdir("relative").is_none());
        assert!(validate_session_tmpdir("/larch-self-edit-missing").is_none());
        let root = TempDir::new().expect("temp");
        let session = root.path().join("claude-implement-parity");
        fs::create_dir_all(&session).expect("session");
        assert!(
            validate_session_tmpdir(&session.to_string_lossy()).is_none(),
            "a session outside the accepted roots is refused"
        );
    }
}
